voteValue input to 2-bigint

arnaucube opened the pull request:
Following spec update from: https://github.com/vocdoni/docs/pull/70

voteValue is the hashed value of the *User* vote, composed by two big
integers:
- The raw user vote is a variable-length array of values and its values
do not need to be checked in the circuit. Furthermore, the values can be
encrypted.
- Since the encoded vote values may not fit into a constant number of
circuit inputs, we calculate a summary of the raw user vote using an
EVM-friendly hash function: `sha256(vote_bytes)`.  The output of the
sha256 hash is slightly larger than the field used in SNARKS, so we
split the hash output (32-bytes) into 2 16-byte arrays, take them as
integers (in little-endian), and use them as circuit inputs.

arnaucube committed:
voteValue input to 2-bigint

Following spec update from: https://github.com/vocdoni/docs/pull/70

voteValue is the hashed value of the *User* vote, composed by two big
integers:
- The raw user vote is a variable-length array of values and its values
do not need to be checked in the circuit. Furthermore, the values can be
encrypted.
- Since the encoded vote values may not fit into a constant number of
circuit inputs, we calculate a summary of the raw user vote using an
EVM-friendly hash function: `sha256(vote_bytes)`.  The output of the
sha256 hash is slightly larger than the field used in SNARKS, so we
split the hash output (32-bytes) into 2 16-byte arrays, take them as
integers (in little-endian), and use them as circuit inputs.

arnaucube committed:
Update /web example

ed255 commented on src/franchise.js:
```diff
@@ -58,7 +59,19 @@ class Voter {
+      .update(voteBuffer)
+      .digest("hex");
+   const voteValue = [
+      BigInt("0x" + voteValueHash.slice(0, 32).match(/.{2}/g).reverse().join("")), // little-endian BigInt representation
```
:woozy_face:

adria0 replied:
🤸‍♂️

arnaucube replied:
is in this room the meeting of js fan club? ^^

adria0 commented on circuits/census.circom:
```diff
@@ -41,7 +41,10 @@ template Census(nLevels) {
+	// voteValue is not operated inside the circuit, assuming that in
+	// Circom an input that is not used will be included in the constraints
+	// system and in the witness
+	signal input voteValue[2];
```
```suggestion
	signal input voteHash[2];
```
I think that `voteHash` is a more accurate naming after the definition of the field semantics

arnaucube replied:
Agree, updated

arnaucube replied:
Done a PR in docs repo to update the var name also in the spec: https://github.com/vocdoni/docs/pull/72

adria0 commented on web/circuit.circom:
```diff
@@ -1,3 +1,3 @@
 include "../circuits/census.circom";
 
-component main = Census(4);
+component main = Census(21);
```
Why changed from `4` to `21`?

arnaucube replied:
Used 21 to do some manual testing in browser, updated back to 4.

adria0 commented on src/franchise.js:
```diff
@@ -58,7 +59,19 @@ class Voter {
 }
 
+function computeVoteValue(voteBuffer) {
+   const voteValueHash = crypto.createHash("sha256")
```
AFAIK, it was not `keccak256` in the spec?

arnaucube replied:
No, `keccak256` is not in the spec, it is with `sha256`. The reasoning behind it is in https://github.com/vocdoni/docs/blob/master/src/architecture/protocol/anonymous-voting/zk-census-proof.md#zkinputs-generation:
>  - Since the encoded vote values may not fit into a constant number of circuit inputs, we calculate a summary of the raw user vote using an EVM-friendly hash function: `sha256(vote_bytes)`.  The output of the sha256 hash is slightly larger than the field used in SNARKS, so we split the hash output (32-bytes) into 2 16-byte arrays, take them as integers (in little-endian), and use them as circuit inputs.
>    - `sha256` hash is used, as if necessary in the future it can be [verified inside](https://github.com/iden3/circomlib/blob/master/circuits/sha256/sha256.circom) the circuit. This usage has two characteristics to keep in mind:
>      - `sha256` is twice as expensive as `keccak256` in terms of gas in EVM, but it is implemented in `circom`, so it can be checked inside a circuit
>      - checking the `sha256` inside a circom circuit is expensive in terms of number of constraints (in the current version of this spec, this is not checked inside the circuit)
>  - example:
>    ```go
>    h := sha256.Sum256(voteBytes) // voteBytes can be the votes array converted to bytes, or the encrypted votes
>    b1 := new(big.Int).SetBytes(swapEndianness(h[:16])) // swap endianness, as golang big int package works in big-endian, and we use little-endian
>    b2 := new(big.Int).SetBytes(swapEndianness(h[16:]))
>    ```
>    And the json input of the `voteValue` for the circuit would be: `"voteValue": [b1, b2]`

Anyway, currently we are not chekcing the hash inside the circuit so does not affect at the final number of constraints, but if at some point in the future is needed we can include the check inside the circuit, as `sha256` is implemented in circomlib. And `sha256` is twice expensive than `keccak256` in terms of gas in EVM, but as we are not using EVM in this use case it is not a problem.

arnaucube committed:
Arrange public inputs order between process & user defined

Arrange public inputs order between process & user defined in order to
easier proof verification inside the Vochain.

Also rename input 'voteValue' to 'voteHash'.

adria0 reviewed (approved):
LGTM
